<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2512]:
import pandas as pd
import psycopg2

In [2513]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

# параметры для подключения к нужной базе данных
DBNAME = 'project_sql' # название базы, к которой нужно подключиться;
USER = 'skillfactory' # имя пользователя в СУБД;
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt' # 
HOST = '84.201.134.129' # адрес, по которому нужно подключиться;
PORT = 5432 # порт, к которому нужно подключиться (по умолчанию равен 5432).

In [2514]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [2515]:
# текст запроса
query_3_1 = f'''SELECT
COUNT(id) as "Всего вакансий в базе"
FROM vacancies
'''

In [2516]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\1114379196.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Всего вакансий в базе
0                  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [2517]:
# текст запроса
query_3_2 = f'''SELECT
COUNT(id) as "Всего работодателей"
FROM employers
'''

In [2518]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\3062646138.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Всего работодателей
0                23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [2519]:
# текст запроса
query_3_3 = f'''SELECT
COUNT(id) as "Всего городов с вакансиями"
FROM areas
'''

In [2520]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\928149064.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Всего городов с вакансиями
0                        1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [2521]:
# текст запроса
query_3_4 = f'''SELECT
COUNT(id) as "Сферы деятельности работодателей"
FROM industries
'''

In [2522]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\2101691108.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Сферы деятельности работодателей
0                               294

***

In [2523]:
# выводы по предварительному анализу данных
total_numbers = f'''SELECT *
FROM 
(SELECT
COUNT(id) as "Вакансии"
FROM vacancies) as vacancies,
(SELECT
COUNT(id) as "Работодатели"
FROM employers) as employers,
(SELECT
COUNT(id) as "Города"
FROM areas) as areas,
(SELECT
COUNT(id) as "Сферы деятельности"
FROM industries) as Industries
'''

df = pd.read_sql_query(total_numbers, connection)
df







C:\Users\asus\AppData\Local\Temp\ipykernel_12460\3999958531.py:18: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вакансии  Работодатели  Города  Сферы деятельности
0     49197         23501    1362                 294

#### В наличии имеются данные о 49147 вакансиях. Вакансии представлены 294 сферами и предоставлены 23501 работодателем из различных 1362 городов.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [2524]:
# текст запроса
query_4_1 = f'''SELECT 
    a.name "Город",
    count(*) "Кол-во вакансий"
FROM areas a 
JOIN vacancies v ON a.id = v.area_id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 20'''

In [2525]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\2457173557.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Город  Кол-во вакансий
0            Москва             5333
1   Санкт-Петербург             2851
2             Минск             2112
3       Новосибирск             2006
4            Алматы             1892
5      Екатеринбург             1698
6   Нижний Новгород             1670
7            Казань             1415
8         Краснодар             1301
9            Самара             1144
10   Ростов-на-Дону             1131
11          Воронеж             1063
12          Ташкент              851
13       Красноярск              847
14       Нур-Султан              840
15        Челябинск              786
16            Пермь              771
17              Уфа              767
18            Томск              692
19             Омск              617

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [2526]:
# текст запроса
query_4_2 = f'''SELECT 
    COUNT(*) as "Вакансий с данными о зарплате"
FROM vacancies
WHERE salary_from is not null or salary_to is not null --фильтр когда одна из колонок с зарплатой с не нулевым значением''' 



In [2527]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df


C:\Users\asus\AppData\Local\Temp\ipykernel_12460\3981930701.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вакансий с данными о зарплате
0                          24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [2528]:
# текст запроса
query_4_3 = f'''SELECT *
FROM -- создадим два отдельных подзапроса и выведем их как одну выборку
(SELECT 
    round(avg(salary_from)) as "Средний минимум"
FROM vacancies) as salary_from,

(SELECT 
    round(avg(salary_to)) as "Средний максимум"
FROM vacancies) as salary_to'''

In [2529]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\2500607024.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Средний минимум  Средний максимум
0          71065.0          110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [2510]:
# текст запроса
query_4_4 = f'''SELECT 
    schedule "График работы",
    employment "Занятость",
    count(*) "Количество"
FROM vacancies
GROUP BY 1, 2
ORDER BY 3 desc'''

In [2530]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
display(df.head(5))

import plotly.express as px
fig = px.pie(df, values='Количество', names = 'График работы', title='Предлагаемый график работы')
fig_1 = px.pie(df, values='Количество', names = 'Занятость', title='Предлагаемая занятость')
fig.show()
fig_1.show()

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\1480368802.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



График работы            Занятость  Количество
0       Полный день     Полная занятость       35367
1  Удаленная работа     Полная занятость        7802
2     Гибкий график     Полная занятость        1593
3  Удаленная работа  Частичная занятость        1312
4    Сменный график     Полная занятость         940

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [2531]:
# текст запроса
query_4_5 = f'''SELECT 
    experience "Опыт работы",
    count(*) "Количество"
FROM vacancies
GROUP BY 1 
ORDER BY 2'''

In [2532]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df
import plotly.express as px
fig = px.pie(df, values='Количество', names='Опыт работы', title='Визуализация требуемого опыта')
fig.show()

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\2448875919.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



***

In [2533]:
# выводы по детальному анализу вакансий

1. Пятерку городов лидеров по количеству вакансий составляют:
    * Москва
    * Питер
    * Минск
    * Новосибирск
    * Алматы
2. Почти половина вакансий из представленной выборки имеет информацию о предоставляемой зарплате. 
3.  71065 - средний показатель нижнего порога зарплаты 
    110537 - средний показатель верхнего порога зарплаты
4.  Большинство работодателей предлагают работу с полной занятостью на полный день. 
Приблизительно 5 часть из всех вакансий предполагают удаленную работу. 
Не значительная часть вакансий предлагают работу с гибким графиком, либо частичной занятостью.
5.  53% из всех вакансий подойдет для начинающего специалиста, 
    30% вакансий предлагает работу для опытных специалистов с опытом от 3 лет,
    14.6% вакансий для соискателей без опыта работы


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [2534]:
# текст запроса
query_5_1 = f'''SELECT 
    e.name,
    COUNT(*)
FROM vacancies v 
JOIN employers e ON v.employer_id = e.id
GROUP BY 1 
ORDER BY 2 DESC
LIMIT 5 
'''

In [2535]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\3834399874.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [2536]:
# текст запроса
query_5_2 = f'''SELECT 
    a.name as "Название города",
    COUNT(distinct e.id) as "кол-во работодателей",
    COUNT(distinct v.id) as "кол-во вакансий"
FROM areas a 
LEFT JOIN employers e ON a.id = e.area 
LEFT JOIN vacancies v ON v.area_id = a.id
WHERE v.id is null
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10
'''

In [2537]:
# результат запроса
df_areas = pd.read_sql_query(query_5_2, connection)
df_areas



C:\Users\asus\AppData\Local\Temp\ipykernel_12460\3394434.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название города  кол-во работодателей  кол-во вакансий
0                 Россия                   410                0
1              Казахстан                   207                0
2     Московская область                    75                0
3     Краснодарский край                    19                0
4               Беларусь                    18                0
5     Ростовская область                    18                0
6            Азербайджан                    17                0
7  Нижегородская область                    16                0
8   Республика Татарстан                    16                0
9             Узбекистан                    15                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [2538]:
# текст запроса
query_5_3 = f'''SELECT 
    e.name "Название работодателя",
    COUNT(DISTINCT area_id) "Кол-во городов"
FROM employers e 
LEFT JOIN vacancies v ON e.id = v.employer_id
GROUP BY 1
ORDER BY 2 DESC

'''

In [2539]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\201667002.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название работодателя  Кол-во городов
0                      Яндекс             181
1                  Ростелеком             152
2                  Спецремонт             116
3      Поляков Денис Иванович              88
4                   ООО ЕФИН               71
...                       ...             ...
23170                  СДЕЛКА               0
23171            Alandr Group               0
23172                     СДК               0
23173      Lemon Land Lombard               0
23174                  Махина               0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [2540]:
# текст запроса
query_5_4 = f'''SELECT 
    COUNT(*) as with_no_industry_info
FROM employers e 
LEFT JOIN employers_industries i ON e.id = i.employer_id
LEFT JOIN industries ind ON ind.id = i.industry_id
WHERE ind.name is null 
'''

In [2541]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\3493105328.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



with_no_industry_info
0                   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [2542]:
# текст запроса
query_5_5 = f'''SELECT 
    e.name,
    Count(*)
FROM employers e 
JOIN employers_industries i ON e.id = i.employer_id
GROUP BY 1

HAVING count(*) = 4
ORDER BY e.name
OFFSET 2
LIMIT 1 
'''

In [2543]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\388711819.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [2544]:
# текст запроса
query_5_6 = f'''SELECT 
    COUNT(*) as "Работодатели связанные с разработкой программного обеспечения"
FROM employers e 
JOIN employers_industries i ON e.id = i.employer_id
JOIN industries ind ON ind.id = i.industry_id
WHERE ind.name ilike '%Разработка программного обеспечения' 
'''

In [2545]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\821628714.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Работодатели связанные с разработ
0                               3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [2546]:
# код для получения списка городов-милионников

In [2547]:
# текст запроса
query_5_7 = f''' 
'''

In [2548]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\1965596796.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



DatabaseError: Execution failed on sql ' 
': can't execute an empty query

***

In [ ]:

#Визулизируем ТОП-10 регинов по кол-ву работодателей
df_areas = pd.read_sql_query(query_5_2, connection)
df_areas

import plotly.express as px
fig = px.pie(df_areas, values='кол-во работодателей', names='Название города', title='ТОП-10 регионов по кол-ву работодателей')
fig.show()

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\692855642.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\692855642.py:5: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



1. Из выборки видно, что лидером среди IT вакансий является Яндекс. При чем, кол-во вакансий у Яндекса больше почти в 4 раза по сравнению с другими компаниями лидерами на рынке.
2. В целом, из данной выборки можно сделать выводы, что работодатели из базы данных зарегистрированы в России, или субъектах Российской федерации и постсоветском пространстве. Лидером является Россия, Казахстан на втором месте, остальные республики представлены малым кол-вом работодателей.
3. 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# текст запроса
query_6_1 = f'''SELECT 
   COUNT(*) as "Кол-во вакансий в Data Science"
FROM vacancies
WHERE lower(name) like '%data%' or lower(name) like '%данн%'
'''


In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\366702258.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Кол-во вакансий в Data Science
0                            1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# текст запроса
query_6_2 = f'''SELECT 
   COUNT(*) as "Вакансий для старта в Data Science"
FROM vacancies
where (experience = 'Нет опыта' or employment = 'Стажировка' or name ilike '%junior%') 
and (name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%')) 
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\2267038862.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вакансий для старта в Data Science
0                                  51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query_6_3 = f'''SELECT 
   COUNT(*) as "Вакансий DS с sql_postgres_skills"
FROM vacancies
where (key_skills ilike '%SQL%' or key_skills ilike '%postgres%' ) 
and (name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%machine learning%' 
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%'))  
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\3794299702.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вакансий DS с sql_postgres_skills
0                                201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query_6_4 = f'''SELECT 
   COUNT(*) as "Вакансий DS с python_skill"
FROM vacancies
where key_skills ilike '%PYTHON%' 
and (name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%machine learning%' 
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%'))
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\4264665179.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вакансий DS с python_skill
0                         351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса
query_6_5 = f'''SELECT
    ROUND(AVG(LENGTH(key_skills)-LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2) cnt_str
FROM vacancies 
WHERE key_skills IS NOT NULL and
(name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%'))  
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\3654210517.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt_str
0     6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# текст запроса
query_6_6 = f'''SELECT 
experience, 
ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0) AS avg_salary
FROM vacancies 
WHERE (salary_from IS not NULL or salary_to IS not NULL) 
AND (name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%'))
AND experience = 'От 3 до 6 лет'
GROUP BY 1
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_12460\519194770.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  avg_salary
0  От 3 до 6 лет    243115.0

***

In [ ]:
# выводы по предметному анализу

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования